# Web Mining and Applied NLP (44-620)

## Requests, JSON, and NLP

### Student Name: Adrianna Webb

[Github Repository](https://github.com/AdriannaWebb/json-sentiment)

Perform the tasks described in the Markdown cells below.  When you have completed the assignment make sure your code cells have all been run (and have output beneath them) and ensure you have committed and pushed ALL of your changes to your assignment repository.

Make sure you have [installed spaCy and its pipeline](https://spacy.io/usage#quickstart) and [spaCyTextBlob](https://spacy.io/universe/project/spacy-textblob)

Every question that requires you to write code will have a code cell underneath it; you may either write your entire solution in that cell or write it in a python file (`.py`), then import and run the appropriate code to answer the question.

This assignment requires that you write additional files (either JSON or pickle files); make sure to submit those files in your repository as well.

In [26]:
import requests
import json
import pickle
import requests
import spacy
from spacytextblob.spacytextblob import SpacyTextBlob

print('All prereqs installed.')
!pip list

All prereqs installed.
Package                   Version
------------------------- --------------
annotated-types           0.7.0
anyio                     4.9.0
argon2-cffi               25.1.0
argon2-cffi-bindings      21.2.0
arrow                     1.3.0
asttokens                 3.0.0
async-lru                 2.0.5
attrs                     25.3.0
babel                     2.17.0
beautifulsoup4            4.13.4
bleach                    6.2.0
blis                      1.3.0
catalogue                 2.0.10
certifi                   2025.7.14
cffi                      1.17.1
charset-normalizer        3.4.2
click                     8.2.1
cloudpathlib              0.21.1
colorama                  0.4.6
comm                      0.2.2
confection                0.1.5
cymem                     2.0.11
debugpy                   1.8.15
decorator                 5.2.1
defusedxml                0.7.1
en_core_web_sm            3.8.0
executing                 2.2.0
fastjsonschema          

### 1. The following code accesses the [lyrics.ovh](https://lyricsovh.docs.apiary.io/#reference/0/lyrics-of-a-song/search) public api, searches for the lyrics of a song, and stores it in a dictionary object.  Write the resulting json to a file (either a JSON file or a pickle file; you choose). You will read in the contents of this file for future questions so we do not need to frequently access the API.

In [27]:
# This API isn't currently working, so I used the alternative API suggested 

#import requests
#import json

#result = json.loads(requests.get('https://api.lyrics.ovh/v1/They Might Be Giants/Birdhouse in your soul').text)

In [28]:
import requests
import json

AUTHOR='Edgar Allan Poe'
POEM = 'A Dream Within A Dream'

#only certain poets and titles are available
#to see the available poets, go to (in a web browser)
# https://poetrydb.org/author
#To see which poems that author has available, go to 
# https://poetrydb.org/author/AUTHOR NAME
# e.g.: https://poetrydb.org/author/Edgar Allan Poe
#The spaces will get handled by your web browser

# A cool pythonism (introduced in Python 3): f strings
# https://docs.python.org/3/tutorial/inputoutput.html#tut-f-strings
URL = f'https://poetrydb.org/author,title/{AUTHOR};{POEM}'
result = json.loads(requests.get(URL).text)
poem = '\n'.join(result[0]['lines']) 

with open("poem_data.json", "w") as f:
    json.dump(result, f, indent=2)

print(poem)   


Take this kiss upon the brow!
And, in parting from you now,
Thus much let me avow--
You are not wrong, who deem
That my days have been a dream:
Yet if hope has flown away
In a night, or in a day,
In a vision or in none,
Is it therefore the less _gone_?
_All_ that we see or seem
Is but a dream within a dream.

I stand amid the roar
Of a surf-tormented shore,
And I hold within my hand
Grains of the golden sand--
How few! yet how they creep
Through my fingers to the deep
While I weep--while I weep!
O God! can I not grasp
Them with a tighter clasp?
O God! can I not save
_One_ from the pitiless wave?
Is _all_ that we see or seem
But a dream within a dream?


### 2. Read in the contents of your file.  Print the lyrics of the song (not the entire dictionary!) and use spaCyTextBlob to perform sentiment analysis on the lyrics.  Print the polarity score of the sentiment analysis.  Given that the range of the polarity score is `[-1.0,1.0]` which corresponds to how positive or negative the text in question is, do you think the lyrics have a more positive or negative connotaion?  Answer this question in a comment in your code cell.

In [29]:
import json
import spacy
from spacytextblob.spacytextblob import SpacyTextBlob

# Read the JSON file from Question 1
with open("poem_data.json", "r") as f:
    poem_data = json.load(f)

# Get just the poem text (not the entire dictionary)
poem_text = '\n'.join(poem_data[0]['lines'])
print("Poem:")
print(poem_text)

# Set up spaCy with sentiment analysis
nlp = spacy.load("en_core_web_sm")
nlp.add_pipe("spacytextblob")

# Perform sentiment analysis on the poem
doc = nlp(poem_text)
polarity_score = doc._.blob.polarity

print(f"\nSentiment Analysis:")
print(f"Polarity score: {polarity_score}")
print(f"Range is [-1.0, 1.0] where -1.0 is most negative and 1.0 is most positive")

# Answer the question about positive/negative connotation
if polarity_score > 0:
    print("The poem has a more POSITIVE connotation.")
elif polarity_score < 0:
    print("The poem has a more NEGATIVE connotation.")
else:
    print("The poem is NEUTRAL.")

Poem:
Take this kiss upon the brow!
And, in parting from you now,
Thus much let me avow--
You are not wrong, who deem
That my days have been a dream:
Yet if hope has flown away
In a night, or in a day,
In a vision or in none,
Is it therefore the less _gone_?
_All_ that we see or seem
Is but a dream within a dream.

I stand amid the roar
Of a surf-tormented shore,
And I hold within my hand
Grains of the golden sand--
How few! yet how they creep
Through my fingers to the deep
While I weep--while I weep!
O God! can I not grasp
Them with a tighter clasp?
O God! can I not save
_One_ from the pitiless wave?
Is _all_ that we see or seem
But a dream within a dream?

Sentiment Analysis:
Polarity score: 0.055555555555555546
Range is [-1.0, 1.0] where -1.0 is most negative and 1.0 is most positive
The poem has a more POSITIVE connotation.


### 3. Write a function that takes an artist, song, and filename, accesses the lyrics.ovh api to get the song lyrics, and writes the results to the specified filename.  Test this function by getting the lyrics to any four songs of your choice and storing them in different files.

In [30]:
import requests
import json

def get_poem_and_save(author, poem_title, filename):
    """
    Function to get poem from Poetry DB API and save to JSON file
    Parameters:
    - author: poet's name
    - poem_title: title of the poem  
    - filename: name of file to save (without .json extension)
    """
    try:
        # Make API request
        url = f'https://poetrydb.org/author,title/{author};{poem_title}'
        response = requests.get(url)
        
        if response.status_code == 200:
            result = response.json()
            
            # Save to file
            with open(f"{filename}.json", "w") as f:
                json.dump(result, f, indent=2)
            
            print(f"Successfully saved '{poem_title}' by {author} to {filename}.json")
            return True
        else:
            print(f"Error {response.status_code} for '{poem_title}' by {author}")
            return False
            
    except Exception as e:
        print(f"Error getting '{poem_title}' by {author}: {e}")
        return False

# Test the function with 4 different poems
poems = [
    ("William Shakespeare", "Sonnet 18", "shakespeare_sonnet"),
    ("Edgar Allan Poe", "Annabel Lee", "poe_annabel"),
    ("Edgar Allan Poe", "The Raven", "raven_poem"),
    ("Edgar Allan Poe", "The Bells", "poe_bells")
]

for author, title, filename in poems:
    get_poem_and_save(author, title, filename)

Successfully saved 'Sonnet 18' by William Shakespeare to shakespeare_sonnet.json
Successfully saved 'Annabel Lee' by Edgar Allan Poe to poe_annabel.json
Successfully saved 'The Raven' by Edgar Allan Poe to raven_poem.json
Successfully saved 'The Bells' by Edgar Allan Poe to poe_bells.json


### 4. Write a function that takes the name of a file that contains song lyrics, loads the file, performs sentiment analysis, and returns the polarity score.  Use this function to print the polarity scores (with the name of the song) of the three files you created in question 3.  Does the reported polarity match your understanding of the song's lyrics? Why or why not do you think that might be?  Answer the questions in either a comment in the code cell or a markdown cell under the code cell.

In [31]:


# Set up spaCy (only need to do this once)
nlp = spacy.load("en_core_web_sm")
nlp.add_pipe("spacytextblob")

def analyze_poem_sentiment(filename):
    """
    Function to load a poem file and return its sentiment polarity score
    """
    try:
        with open(filename, "r") as f:
            poem_data = json.load(f)
        
        poem_text = '\n'.join(poem_data[0]['lines'])
        
        doc = nlp(poem_text)
        polarity_score = doc._.blob.polarity
        
        return polarity_score
        
    except Exception as e:
        print(f"Error analyzing {filename}: {e}")
        return None

# Test the function on the files created in Question 3
files_to_analyze = ["shakespeare_sonnet.json", "poe_annabel.json", "raven_poem.json", "poe_bells.json"]

print("Sentiment Analysis Results:")

for filename in files_to_analyze:
    polarity = analyze_poem_sentiment(filename)
    if polarity is not None:
        print(f"{filename}: {polarity:.3f}")
        
        # Interpret the score
        if polarity > 0.1:
            sentiment = "Positive"
        elif polarity < -0.1:
            sentiment = "Negative"  
        else:
            sentiment = "Neutral"
        print(f"   {sentiment} sentiment")
        print()

Sentiment Analysis Results:
shakespeare_sonnet.json: 0.332
   Positive sentiment

poe_annabel.json: 0.360
   Positive sentiment

raven_poem.json: 0.038
   Neutral sentiment

poe_bells.json: 0.081
   Neutral sentiment



### 4.1 Does the reported polarity match your understanding of the song's lyrics?

I think the Shakespear sonnet is more positive than 0.332 but I can understand why the sentiment score was lower because of the negative connotation of some of the words used. Other than that I think that the score is accurate if not a little to high for Raven. 

